src/mam4xx/wet_dep.hpp

```diff
@@ -0,0 +1,178 @@
+ *      as outer loops will iterate over columns, so we drop ncol as input.
+ * @pre In F90 code, pcols is the number of columns in the mesh.
+ *      Since we are only operating over one column, pcols == 1.
+ * @pre In F90 code, ncol == pcols. Since ncol == 1, pcols == 1.
+ *
+ * @pre atm is initialized correctly and has the correct number of levels.
+ * 
+ */
+// clang-format on 
+KOKKOS_INLINE_FUNCTION
+void calculate_cloudy_volume(/* cont int ncol, */ const Real *cld, const Real *lprec,
+                             const bool is_tot_cld, Real *cldv, Real *sumppr_all,
+                             const Atmosphere &atm) {
+  const int pver = atm.num_levels();
+  Real sumppr = 0.0; // Precipitation rate [kg/m2/s]
+  Real cldv1 = 0.0; // Precip weighted cloud fraction from above [kg/m2/s]
+  Real sumpppr = 1e-36; // Sum of positive precips from above
+
+  Real lprecp = 0.0; // Local production rate of precip [kg/m2/s] if positive
+
+  for (int i = 0; i < pver; i++) {
+      if (is_tot_cld) {
+          cldv[i] = haero::max( haero::min(1.0, cldv1 / sumpppr) * sumppr / sumpppr, cld[i]);
+      }
+      else {
+          // For convective and stratiform precipitation volume at the top 
+          // interface of each layer. Neglect the current layer.
+          cldv[i] = haero::max( haero::min(1.0, cldv1 / sumpppr) * (sumppr / sumpppr), 0.0);
+      }
+      lprecp = haero::max(lprec[i], 1e-30);
+      cldv1 += cld[i] * lprecp;
+      sumppr += lprec[i];
+      sumppr_all[i] = sumppr;
+      sumpppr += lprecp;
+  }
+}
+
+/**
+ * @brief Calculate rain mixing ratio [kg/kg] from the local precipitation rate
+ * 
+ * @param[in] temperature Temperature [K].
+ * @param[in] pmid Pressure at layer midpoints [Pa].
+ * @param[in] sumppr Sum of precipitation rate above each layer [kg/m2/s]
+ * @param[in] atm Atmosphere object (used for number of levels).
+ * 
+ * @param[out] rain Rain mixing ratio [kg/kg].
+ * 
+ * @pre In F90, ncol == 1 as we only operate over one column at a time
+ *      as outer loops will iterate over columns, so we drop ncol as input.
+ * @pre In F90 code, pcols is the number of columns in the mesh.
+ *      Since we are only operating over one column, pcols == 1.
+ * @pre In F90 code, ncol == pcols. Since ncol == 1, pcols == 1.
+ *
+ * @pre atm is initialized correctly and has the correct number of levels.
+*/
+KOKKOS_INLINE_FUNCTION
+void rain_mix_ratio(/* cont int ncol, */ const Real *temperature, const Real *pmid,
+                             const Real *sumppr, Real *rain, const Atmosphere &atm) {
+  const int pver = atm.num_levels();
+  Real convfw = 0.0; // Falling velocity at air density = 1 kg/m3 [m/s * sqrt(rho)].
+  Real rho = 0.0; // Air density [kg/m3].
+  Real vfall = 0.0; // Falling velocity [m/s].
+
+  // Define constant convfw from cldwat.F90
+  // Reference: Tripoli and Cotton (1980)
+  convfw = 1.94 * 2.13 * haero::sqrt(Constants::density_h2o * Constants::gravity * 2.7e-4);
```

Comment:
Not all constants exist at the Haero ("generic aerosol package") level, since different aerosol packages use different physical assumptions. In other words, we're probably not going to have a single namespace with all the physical constants in it.

Reply:
huh - so should I avoid using these constants then? Are they subject to change in the future and should be redefined at the class level?

Reply:
You can use any constants that exist in that namespace. If you don't find them there, you'll have to create them or use them at the mam4xx level.